### DISCUSSION TITLE
Adding new translation services by users themselves

### DISCUSSION THREAD

**Discussion by iG8R** [immersive-translate/immersive-translate#936] [Ideas]
### 新功能 | New Feature

Hi.
Is there any way to implement a feature such as adding new translation services by users themselves when the preparation of the original text on the page (breakdown into paragraphs by number of characters, etc.) for transmission to and distribution of the acquired translated text on the page itself from the translation server will fall on the shoulders of ImmersiveTranslate, while the formation of the request itself to and processing of the received result from the translation server will be done by the user code similar to the implementation on QTranslate. 
Here is a description of QTranslate: https://www.52pojie.cn/thread-1695613-1-1.html

Here is an example of how the request to and processing of the response from Google Translate takes place.
<details><summary>GoogleTranslate user code</summary>

```
function serviceHeader() {
    return new ServiceHeader(1, "Google", "Google's free online language translation service instantly translates text and web pages." + Const.NL2 + serviceHost() + Const.NL2 + "\u00a9 2020 Google", Capability.TRANSLATE | Capability.DETECT_LANGUAGE | Capability.LISTEN)
}
function serviceHost(a, b, c) {
    return "https://translate.google." + (Options.PreferredDomain || Options.GoogleDomain || "com")
}
function serviceLink(a, b, c) {
    var h = serviceHost() + "/";
    a && (b = isLanguage(b) ? codeFromLanguage(b) : "auto", c = isLanguage(c) ? codeFromLanguage(c) : "auto", h += format("#{0}/{1}/{2}", b, c, encodeGetParam(a)));
    return h
}
SupportedLanguages = [-1, "auto", "af", "az", "sq", "ar", "hy", "eu", "be", "bg", "ca", "zh-CN", "zh-TW", "hr", "cs", "da", "nl", "en", "et", "fi", "tl", "fr", "gl", "de", "el", "ht", "iw", "hi", "hu", "is", "id", "it", "ga", "ja", "ka", "ko", "lv", "lt", "mk", "ms", "mt", "no", "fa", "pl", "pt", "ro", "ru", "sr", "sk", "sl", "es", "sw", "sv", "th", "tr", "uk", "ur", "vi", "cy", "yi", "eo", "hmn", "la", "lo", "kk", "uz", "si", "tg", "te", "km", "mn", "kn", "ta", "mr", "bn", "tt"];
function tk(a) {
    function b(a, b) {
        for (var d = 0; d < b.length - 2; d += 3) {
            var c = b.charAt(d + 2),
            c = "a" <= c ? c.charCodeAt(0) - 87 : Number(c),
            c = "+" == b.charAt(d + 1) ? a >>> c : a << c;
            a = "+" == b.charAt(d) ? a + c & 4294967295 : a ^ c
        }
        return a
    }
    for (var c = (Options.GoogleTkk || "0.0").split("."), h = Number(c[0]) || 0, g = [], d = 0, e = 0; e < a.length; e++) {
        var f = a.charCodeAt(e);
        128 > f ? g[d++] = f : (2048 > f ? g[d++] = f >> 6 | 192 : (55296 == (f & 64512) && e + 1 < a.length && 56320 == (a.charCodeAt(e + 1) & 64512) ? (f = 65536 + ((f & 1023) << 10) + (a.charCodeAt(++e) & 1023), g[d++] = f >> 18 | 240, g[d++] = f >> 12 & 63 |
                                128) : g[d++] = f >> 12 | 224, g[d++] = f >> 6 & 63 | 128), g[d++] = f & 63 | 128)
    }
    a = h;
    for (d = 0; d < g.length; d++)
        a += g[d], a = b(a, "+-a^+6");
    a = b(a, "+-3^+b+-f");
    a ^= Number(c[1]) || 0;
    0 > a && (a = (a & 2147483647) + 2147483648);
    a %= 1E6;
    return a.toString() + "." + (a ^ h)
}
function serviceDetectLanguageRequest(a) {
    a = limitSource(a);
    var b = tk(a);
    a = encodeUriParam(a);
    var c = a.length > Const.MAX_URI_LEN ? HttpMethod.POST : HttpMethod.GET;
    return new RequestData(c, "/translate_a/single?client=gtx&sl=auto&dt=ld&ie=UTF-8&oe=UTF-8&tk=" + b + (c === HttpMethod.POST ? "" : "&q=" + a), c === HttpMethod.POST ? "q=" + a : null)
}
function serviceDetectLanguageResponse(a) {
    a = parseJSON(a);
    return getSourceLanguage(a)
}
function serviceTranslateRequest(a, b, c) {
    a = limitSource(a);
    var h = tk(a);
    a = encodeUriParam(a);
    var g = a.length > Const.MAX_URI_LEN ? HttpMethod.POST : HttpMethod.GET;
    b = format("/translate_a/single?client=gtx&sl={0}&tl={1}&hl={2}&dt=bd&dt=t&dt=ld&dt=rm&ie=UTF-8&oe=UTF-8&tk={3}", codeFromLanguage(b), codeFromLanguage(c), Options.LanguageCode, h);
    g === HttpMethod.GET && (b += "&q=" + a);
    return new RequestData(g, b, g === HttpMethod.POST ? "q=" + a : null)
}
function serviceTranslateResponse(a, b, c, h) {
    a = parseJSON(b);
    var g = b = "",
    d,
    e;
    if (a) {
        var f = a[0];
        if (f)
            for (d = 0; d < f.length; d++)
                (e = f[d]) && 0 !== e.length && (b += e[0] || "", 2 < e.length && d === f.length - 1 && (g = e[2] || ""));
        if (f = a[1])
            for (d = 0; d < f.length; d++)
                if (e = f[d], !(3 > e.length)) {
                    b += Const.NL2 + e[0] + ":";
                    var l = e[2];
                    for (e = 0; e < l.length; e++) {
                        var k = l[e];
                        1 < k.length && (b += Const.NL + "    " + k[0], (k = k[1]) && 0 < k.length && (b += " (" + k.join(", ") + ")"))
                    }
                }
        isLanguage(c) || (c = getSourceLanguage(a))
    }
    return new ResponseData(b, c, h, g)
}
function getSourceLanguage(a) {
    return !a || 9 > a.length ? UNKNOWN_LANGUAGE : (a = a[8]) && 0 !== a.length ? languageFromCode(a[0][0]) : UNKNOWN_LANGUAGE
}
function serviceListenRequest(a, b, c) {
    a = format("/translate_tts?ie=UTF-8&q={0}&tl={1}&client=gtx&tk={2}", encodeGetParam(a), codeFromLanguage(b), tk(a));
    c && (a += "&ttsspeed=0.24");
    return new RequestData(HttpMethod.GET, a)
};

```
</details>

### 为什么您想要添加这些功能 | Why you want to add these features

Today, more and more new translation services are appearing, including intracorporate ones. This system will make it much easier to add such services to ImmersiveTranslate without the need to intervene in the add-on's code itself.

**Comment by iG8R** [comment]
To be more precise.

Here are the requests and responses from the private intracorporate ChatGPT server.
![ImmersiveTranslate-ownChatGPTserver-Headers03-2023-09-04_200146](https://github.com/immersive-translate/immersive-translate/assets/11407417/13093e02-4b7e-4950-84fa-c13999000da5)
![ImmersiveTranslate-ownChatGPTserver-Request-2023-09-04_195719](https://github.com/immersive-translate/immersive-translate/assets/11407417/18cfcdeb-25be-4b7c-857e-ef69e93fe48d)
![ImmersiveTranslate-ownChatGPTserver-Response02-2023-09-04_195745](https://github.com/immersive-translate/immersive-translate/assets/11407417/5a1ba4e0-aefb-477b-be91-b7f9393057ab)

And here is a user code for Qtranslate to process both request and response from this server.

```
function serviceHeader() {
    return new ServiceHeader(108, "ChatGPT", "Artificial intelligence OpenAI as a professional translator." + Const.NL2 + "https://www.companydomain02.com/" + Const.NL2 + "\u00a9 OpenAI", Capability.TRANSLATE)
}

function serviceHost(a, b, c) {
    return "https://api.companydomain01.com"
}

SupportedLanguages = [-1, -1, "af","az","sq","ar","hy","eu","be","bg","ca","zh-CN","zh-TW","hr","cs","da","nl","en","et","fi","tl","fr","gl","de","el","ht","iw","hi","hu","is","id","it","ga","ja","ka","ko","lv","lt","mk","ms","mt","no","fa","pl","pt","ro","ru","sr","sk","sl","es","sw","sv","th","tr","uk","ur","vi","cy","yi","eo","hmn","la","lo","kk","uz","si","tg","te","km","mn","kn","ta","mr","bn","tt"];

function serviceTranslateRequest(a, b, c) {
	
   return new RequestData(
		HttpMethod.POST, 
		"", 
		stringifyJSON(
			{
				"q":"Translate from "+codeFromLanguage(b || AUTO_DETECT_LANGUAGE) +" to "+ codeFromLanguage(c || ENGLISH_LANGUAGE)+" while preserving the formatting of the text:"+Const.NL+ a,
				"type":"text_translation"
			}
		), postHeader("application/json")+Const.NL+
		"cache-control: no-cache"+Const.NL+
		"pragma: no-cache"+Const.NL+
		"sec-ch-ua: \"Not/A)Brand\";v=\"99\", \"Google Chrome\";v=\"115\", \"Chromium\";v=\"115\""+Const.NL+
		"sec-ch-ua-mobile: ?0"+Const.NL+
		"sec-ch-ua-platform: \"Windows\""+Const.NL+
		"sec-fetch-dest: empty"+Const.NL+
		"sec-fetch-mode: cors"+Const.NL+
		"sec-fetch-site: cross-site"+Const.NL+
		"Referer: https://www.companydomain02.com/"+Const.NL+
		"Referrer-Policy: strict-origin-when-cross-origin"
		)
}

function serviceTranslateResponse(a, b, c, d) {
	ff = parseJSON(b);
	if (ff.response.choices[0].message.content){
		b = ff.response.choices[0].message.content;
		b = b.replace(/^\n+/gm, "");
	}
    return new ResponseData(b, c, d)
};
```

**Comment by theowenyoung** [comment]
We prefer another way for this custom API, and we have supported it now aon Beta feature, you should enable it on developer settings.

<img width="1169" alt="image" src="https://github.com/immersive-translate/immersive-translate/assets/62473795/e3839104-8ef0-4249-9d89-a97050442dab">

Once it enabled, you can see a translation service: Custom API:

<img width="886" alt="image" src="https://github.com/immersive-translate/immersive-translate/assets/62473795/dad74fe8-579d-40e3-8cd1-3000dd96ee5e" />

 You'll need to implement an API on your server, or locally to you, that follows our request and response format.

 The specific documentation is here: https://immersivetranslate.com/docs/services/custom/

**Comment by iG8R** [comment]
The point is that, with such a custom API support system, it is necessary to make changes on the server itself.
But what if there is no access to the server?
But with the system that I mentioned above there won't be any issues at all, which is confirmed by the fact that access to the intracorporate translation server was easily obtained with QTranslate without any need for server-side changes.

As an example, the author of QTranslate has already abandoned this program for about 2 years, but it is still alive and working because it allows to easily fix the problem caused by changes on the translation servers' side.
ImmersiveTranslate is a great addon, it's just that there are always concerns that there may come a time when a situation similar to QTranslate arises and it stops working :(

**Comment by theowenyoung** [comment]
I understand what you're saying, you want immersive translation to allow dynamic code injection, kind of like Grease Monkey Scripts, but in the manifest v3 version we are using now, Chrome no longer allows us to insert dynamic code: https://developer.chrome.com/docs/extensions/ migrating/improve-security/#remove-execution-of-strings , the Grease Monkey scripts use manifest v2 , so they can still use the dynamic code.

Another reason is that we have our own security concerns, and since we provide a user system, Pro Membership, we don't really want to be able to execute user code on our system either!

**Comment by iG8R** [comment]
Oh, I understand regarding security concerns, but for those, who knowledgeable in JS, would it be a problem to tamper ImmersiveTranslate code for their dirty deeds if they want something bad to do?

Is it possible to somehow make it so that in ImmersiveTranslate there is no need to inject foreign code, but only have the ability to specify the necessary fields of the JSON request and Referer for a specific server, as well as the name of the response field from the server, in which the translated text is passed?

**Reply by theowenyoung** [reply]
Theoretically it can be done, but designing such a static configuration is still a bit of a challenge.

@keyboard3  has tried to design this before, but we've found it difficult to cater for a variety of scenarios. @keyboard3  can you help us see if the requests and responses here can be expressed in the static configuration you have designed before?
